File: ios/Reguerta/Reguerta/Data/Media/ImageUploadFileNameFormatter.swift
import Foundation

enum ImageUploadFileNameFormatter {
    static func formatPrefix(
        nameHint: String?,
        namespace: ImageUploadNamespace,
        maxLength: Int = defaultPrefixMaxLength
    ) -> String {
        let fallback: String = switch namespace {
        case .products: "product"
        case .news: "news"
        case .sharedProfiles: "profile"
        }

        let normalized = normalizeHint(nameHint)
        let clampedLength = max(1, maxLength)
        let limited = String(normalized.prefix(clampedLength))
        return limited.isEmpty ? fallback : limited
    }

    private static func normalizeHint(_ hint: String?) -> String {
        let source = (hint ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !source.isEmpty else { return "" }

        let deAccented = source.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
        let wordsWithUnderscores = deAccented
            .replacingOccurrences(of: "\\s+", with: "_", options: .regularExpression)
        let alphanumericUnderscoreOnly = wordsWithUnderscores
            .replacingOccurrences(of: "[^A-Za-z0-9_]", with: "", options: .regularExpression)
            .replacingOccurrences(of: "_+", with: "_", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "_"))

        return alphanumericUnderscoreOnly.lowercased()
    }

    private static let defaultPrefixMaxLength = 16
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/media/ImageUploadFileNameFormatter.kt
package com.reguerta.user.data.media

import java.text.Normalizer

internal object ImageUploadFileNameFormatter {
    fun formatPrefix(
        nameHint: String?,
        namespace: ImageUploadNamespace,
        maxLength: Int = DefaultPrefixMaxLength,
    ): String {
        val fallback = when (namespace) {
            ImageUploadNamespace.PRODUCTS -> "product"
            ImageUploadNamespace.NEWS -> "news"
            ImageUploadNamespace.SHARED_PROFILES -> "profile"
        }
        val normalized = normalizeHint(nameHint)
        return normalized
            .take(maxLength.coerceAtLeast(1))
            .ifBlank { fallback }
    }

    private fun normalizeHint(nameHint: String?): String {
        val deAccented = Normalizer.normalize(nameHint.orEmpty(), Normalizer.Form.NFD)
            .replace(Regex("\\p{Mn}+"), "")
        return deAccented
            .trim()
            .replace(Regex("\\s+"), "_")
            .replace(Regex("[^A-Za-z0-9_]"), "")
            .replace(Regex("_+"), "_")
            .trim('_')
            .lowercase()
    }

    private const val DefaultPrefixMaxLength = 16
}
